// react-web-share/dist/react-web-share.esm.js
import React, { useEffect, useState, useCallback, useMemo, cloneElement } from 'react';
import { createPortal } from 'react-dom';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var BackdropContainerStyle = {
  position: "fixed",
  left: 0,
  top: 0,
  width: "100%",
  height: "100%",
  background: "rgba(0, 0, 0, 0.4)",
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "flex-end",
  zIndex: 1400
};
var inEffect = ".web-share-fade{animation:simpleFade 0.5s;animation-fill-mode:both }@keyframes simpleFade{0%{opacity:0 }100%{opacity:1 }}.web-share-fade-in-up{animation:fadeInUp 0.5s;animation-fill-mode:both }@keyframes fadeInUp{0%{opacity:0;transform:translateY(20px) }100%{opacity:1;transform:translateY(0) }}";
function Backdrop(_ref) {
  var children = _ref.children,
      onClose = _ref.onClose;

  var handleOnClose = function handleOnClose(e) {
    if (e.target === e.currentTarget) {
      onClose(e);
    }
  };

  return React.createElement("div", {
    onClick: handleOnClose,
    className: "web-share-fade",
    style: BackdropContainerStyle
  }, React.createElement("style", {
    children: inEffect
  }), children);
}

var externalOpen = function externalOpen(URL) {
  return window.open(URL, "_blank", "noopener");
};

var iconList = {
  facebook: {
    path: /*#__PURE__*/React.createElement("path", {
      d: "M24 12a12 12 0 10-13.9 11.9v-8.4h-3V12h3V9.4c0-3 1.8-4.7 4.6-4.7l2.6.2v3h-1.5c-1.5 0-2 .9-2 1.8V12h3.4l-.5 3.5h-2.8v8.4A12 12 0 0024 12z"
    }),
    color: "#0076FB",
    e: function e(l) {
      return externalOpen("https://www.facebook.com/sharer/sharer.php?u=" + l);
    }
  },
  twitter: {
    path: /*#__PURE__*/React.createElement("path", {
      d: "M24 4.6a10 10 0 01-2.9.7 5 5 0 002.2-2.7c-1 .6-2 1-3.1 1.2a5 5 0 00-8.4 4.5A14 14 0 011.6 3.2 4.8 4.8 0 001 5.6a5 5 0 002.2 4.1 4.9 4.9 0 01-2.3-.6A5 5 0 005 14a5 5 0 01-2.2 0 5 5 0 004.6 3.5 9.9 9.9 0 01-6.1 2.1H0a14 14 0 007.6 2.2c9 0 14-7.5 14-14V7A10 10 0 0024 4.6z"
    }),
    color: "#1DA1F2",
    e: function e(l, t) {
      return externalOpen("https://twitter.com/intent/tweet?text=" + t + "&url=" + l);
    }
  },
  whatsapp: {
    path: /*#__PURE__*/React.createElement("path", {
      d: "M17.5 14.4l-2-1c-.3 0-.5-.1-.7.2l-1 1.1c-.1.2-.3.3-.6.1s-1.3-.5-2.4-1.5a9 9 0 01-1.7-2c-.1-.3 0-.5.2-.6l.4-.6c.2-.1.2-.3.3-.5v-.5L9 7c-.2-.6-.4-.5-.6-.5h-.6c-.2 0-.5 0-.8.4-.2.3-1 1-1 2.5s1 2.8 1.2 3c.2.2 2.1 3.2 5.1 4.5l1.7.6a4 4 0 001.9.2c.5-.1 1.7-.8 2-1.5.2-.6.2-1.2.1-1.4l-.5-.3M12 21.8a9.9 9.9 0 01-5-1.4l-.4-.2-3.7 1 1-3.7-.2-.3a9.9 9.9 0 01-1.5-5.3 9.9 9.9 0 0116.8-7 9.8 9.8 0 013 7 9.9 9.9 0 01-10 9.9m8.4-18.3A11.8 11.8 0 0012.1 0 12 12 0 001.8 17.8L0 24l6.4-1.6a11.9 11.9 0 005.6 1.4 12 12 0 0012-11.9 11.8 11.8 0 00-3.5-8.4z"
    }),
    color: "#25D366",
    e: function e(l, t) {
      return externalOpen("https://api.whatsapp.com/send?text=" + t + " " + l);
    }
  },
  reddit: {
    path: /*#__PURE__*/React.createElement("path", {
      d: "M12 0A12 12 0 000 12a12 12 0 0012 12 12 12 0 0012-12A12 12 0 0012 0zm5.01 4.74c.69 0 1.25.56 1.25 1.25a1.25 1.25 0 01-2.5.06l-2.6-.55-.8 3.75c1.83.07 3.48.63 4.68 1.49.3-.31.73-.5 1.2-.5.97 0 1.76.8 1.76 1.76 0 .72-.43 1.33-1.01 1.61a3.11 3.11 0 01.04.52c0 2.7-3.13 4.87-7 4.87-3.88 0-7-2.17-7-4.87 0-.18 0-.36.04-.53A1.75 1.75 0 014.03 12a1.75 1.75 0 012.96-1.26 8.52 8.52 0 014.74-1.5l.89-4.17a.34.34 0 01.14-.2.35.35 0 01.24-.04l2.9.62a1.21 1.21 0 011.11-.7zM9.25 12a1.25 1.25 0 101.25 1.25c0-.69-.56-1.25-1.25-1.25zm5.5 0a1.25 1.25 0 000 2.5 1.25 1.25 0 000-2.5zm-5.47 3.99a.33.33 0 00-.23.1.33.33 0 000 .46c.84.84 2.49.91 2.96.91.48 0 2.1-.06 2.96-.91a.36.36 0 00.03-.47.33.33 0 00-.46 0c-.55.54-1.68.73-2.51.73-.83 0-1.98-.2-2.51-.73a.33.33 0 00-.24-.1z"
    }),
    color: "#FF4500",
    e: function e(l, t) {
      return externalOpen("https://www.reddit.com/submit?url=" + l + "&title=" + t);
    }
  },
  telegram: {
    path: /*#__PURE__*/React.createElement("path", {
      d: "M23.91 3.79L20.3 20.84c-.25 1.21-.98 1.5-2 .94l-5.5-4.07-2.66 2.57c-.3.3-.55.56-1.1.56-.72 0-.6-.27-.84-.95L6.3 13.7.85 12c-1.18-.35-1.19-1.16.26-1.75l21.26-8.2c.97-.43 1.9.24 1.53 1.73z"
    }),
    color: "#0088CC",
    e: function e(l, t) {
      return externalOpen("https://telegram.me/share/msg?url=" + l + "&text=" + t);
    }
  },
  linkedin: {
    path: /*#__PURE__*/React.createElement("path", {
      d: "M6.5 21.5h-5v-13h5v13zM4 6.5C2.5 6.5 1.5 5.3 1.5 4s1-2.4 2.5-2.4c1.6 0 2.5 1 2.6 2.5 0 1.4-1 2.5-2.6 2.5zm11.5 6c-1 0-2 1-2 2v7h-5v-13h5V10s1.6-1.5 4-1.5c3 0 5 2.2 5 6.3v6.7h-5v-7c0-1-1-2-2-2z"
    }),
    color: "#0073b1",
    e: function e(l, t, ti) {
      return externalOpen("https://www.linkedin.com/shareArticle?mini=true&url=" + l + "&title=" + ti + "&summary=" + t);
    }
  },
  mail: {
    path: /*#__PURE__*/React.createElement("path", {
      d: "M20 4H4a2 2 0 00-2 2v12c0 1.1.9 2 2 2h16a2 2 0 002-2V6a2 2 0 00-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
    }),
    color: "#E53E3E",
    e: function e(l, t) {
      return externalOpen("mailto:?body=" + l + "&subject=" + t);
    }
  },
  copy: {
    path: /*#__PURE__*/React.createElement("path", {
      d: "M16 1H4a2 2 0 00-2 2v14h2V3h12V1zm3 4H8a2 2 0 00-2 2v14c0 1.1.9 2 2 2h11a2 2 0 002-2V7a2 2 0 00-2-2zm0 16H8V7h11v14z"
    }),
    color: "#718096",
    e: function e(l) {
      return navigator.clipboard.writeText(decodeURIComponent(l));
    }
  }
};

var PortalComponent = function PortalComponent(_ref) {
  var children = _ref.children;
  var el = document.createElement("div");
  useEffect(function () {
    document.body.appendChild(el);
    return function () {
      document.body.removeChild(el);
    };
  }, [el]);
  return createPortal(children, el);
};

var CloseStyle = {
  background: "#edf2f7",
  cursor: "pointer",
  padding: "0.75rem",
  display: "block",
  width: "100%",
  border: 0,
  fontSize: "1rem"
};
function CloseButton(_ref) {
  var onClose = _ref.onClose,
      closeText = _ref.closeText;
  return React.createElement("button", {
    style: CloseStyle,
    "aria-label": "Close",
    type: "button",
    onClick: onClose
  }, closeText || "Close");
}

var ShareHeader = {
  padding: "1rem 1.25rem",
  fontSize: "1.25rem",
  fontWeight: 600,
  paddingBottom: 0
};
function Header(_ref) {
  var title = _ref.title;
  return React.createElement("header", {
    style: ShareHeader
  }, title);
}

var SocialIconStyle = {
  width: "100%",
  height: "auto",
  cursor: "pointer",
  border: 0,
  background: "#1A78F6",
  padding: "0.75rem",
  borderRadius: "0.5rem",
  fontSize: 0
};
function Icon(_ref) {
  var name = _ref.name,
      data = _ref.data,
      onClose = _ref.onClose;
  var _iconList$name = iconList[name],
      path = _iconList$name.path,
      _iconList$name$viewBo = _iconList$name.viewBox,
      viewBox = _iconList$name$viewBo === void 0 ? "0 0 24 24" : _iconList$name$viewBo,
      color = _iconList$name.color,
      e = _iconList$name.e;
  return React.createElement("button", {
    onClick: function onClick() {
      return e(encodeURIComponent(data.url), data.text, data.title) && onClose();
    },
    "aria-label": name,
    style: _extends({}, SocialIconStyle, {
      background: color
    })
  }, React.createElement("svg", {
    fill: "white",
    viewBox: viewBox
  }, path));
}

var IconsContainer = {
  maxWidth: "24rem",
  width: "90%",
  backgroundColor: "white",
  borderRadius: "0.5rem",
  borderBottomLeftRadius: 0,
  borderBottomRightRadius: 0
};
var IconsLisStyle = {
  padding: "1.25rem",
  display: "grid",
  gridTemplateColumns: "repeat(4, 1fr)",
  gridGap: "1.25rem"
};
function SocialIcons(_ref) {
  var onClose = _ref.onClose,
      sites = _ref.sites,
      data = _ref.data,
      closeText = _ref.closeText;
  return React.createElement("section", {
    role: "dialog",
    "aria-modal": "true",
    className: "web-share-fade-in-up",
    style: IconsContainer
  }, React.createElement(Header, {
    title: data.title
  }), React.createElement("div", {
    style: IconsLisStyle
  }, sites.map(function (name) {
    return React.createElement(Icon, {
      name: name,
      key: name,
      data: data,
      onClose: onClose
    });
  })), React.createElement(CloseButton, {
    onClose: onClose,
    closeText: closeText
  }));
}

function useDisclosure(isOpenDefault) {
  if (isOpenDefault === void 0) {
    isOpenDefault = false;
  }

  var _useState = useState(isOpenDefault),
      isOpen = _useState[0],
      setIsOpen = _useState[1];

  var onOpen = useCallback(function () {
    return setIsOpen(true);
  }, []);
  var onClose = useCallback(function () {
    return setIsOpen(false);
  }, []);
  var onToggle = useCallback(function () {
    return setIsOpen(function (state) {
      return !state;
    });
  }, []);
  return {
    isOpen: isOpen,
    onOpen: onOpen,
    onClose: onClose,
    onToggle: onToggle
  };
}

var RWebShare = function RWebShare(_ref) {
  var children = _ref.children,
      closeText = _ref.closeText,
      data = _ref.data,
      _ref$sites = _ref.sites,
      sites = _ref$sites === void 0 ? Object.keys(iconList) : _ref$sites;

  var _useDisclosure = useDisclosure(),
      onOpen = _useDisclosure.onOpen,
      onClose = _useDisclosure.onClose,
      isOpen = _useDisclosure.isOpen;

  var shareData = useMemo(function () {
    return _extends({}, data, {
      title: data.title || "share",
      text: data.text || "",
      url: data.url || typeof window !== "undefined" && window.location.href || ""
    });
  }, [data]);

  var handleOnClick = function handleOnClick() {
    try {
      window.navigator.share(shareData);
    } catch (e) {
      console.warn(e);
      onOpen();
    }
  };

  return React.createElement(React.Fragment, null, cloneElement(children, _extends({}, children == null ? void 0 : children.props, {
    onClick: handleOnClick
  })), isOpen && React.createElement(PortalComponent, null, React.createElement(Backdrop, {
    onClose: onClose
  }, React.createElement(SocialIcons, {
    onClose: onClose,
    sites: sites,
    data: shareData,
    closeText: closeText
  }))));
};

export { RWebShare };
